### ISSUE TITLE
skim S3 method for vectors

### ISSUE THREAD

**Issue by jacob-long** [ropenscilabs/skimr#164]
It's very possible this has been considered or is obviously impractical given the way the package is designed; if so, feel free to ignore/flame/etc.

As someone who just downloaded the package to work with some survey data, I knew I didn't want to use `skim` on the data frame because it has hundreds of columns, many of them somewhat redundant or of no interest. 

My intuition during interactive data exploration was to do this: `skim(df$variable)`, but `skim` only accepts data frames. Of course, I get exactly what I want by just doing modifying the call like this: `skim(df["variable"])`, but it's certainly a bit clunkier (and novice users will probably never get that far in asking why it didn't work at first). I know you have `skim_v`, but from what I can tell it doesn't do the pretty printing that `skim` does. 

While I put S3 method in the title, it may be as simple as adding something like

`if (is.vector(.data)) {as.data.frame(.data)} `

to another `skim` method.

**Comment by elinw** [comment]
@michaelquinn32 and I were talking about this the other day, specifically is it better to 
- have the user do their own coercion (`skim(as.data.frame(mtcars$mpg))
- we build up skim.vector, skim.matrix etc or 
- build out a print method for skim_v (you can see that I posted an issue for that idea).

I'd definitely be interested in hearing thoughts on this. 

Since skim() is piped I also think that using `dplyr::select()` is a good solution for when you have a lot of survey data etc.  Part of the process we imagined at the #unconf is that you get a data set, skim it once, then on the basis of that  modify it in some way (like selecting columns, dropping variables that are all NA, etc) and then skimming again in a somewhat iterative process.

**Comment by michaelquinn32** [comment]
We could probably support something like this by integrating tidyselect with the package.
https://github.com/tidyverse/tidyselect

In that approach, you could call

skim(df, whatever_vector_you_want, plus_another)

Unfortunately, skim works on assumptions similar to dplyr, which also doesn't have semantics for calling on individual vectors. skim_v sort of gets you there, but as you point out, skim_v is really designed to do internal work.

**Comment by elinw** [comment]
That seems like a good approach to me and it does match what tidyverse people are used to in terms of being able to list the columns.

**Comment by jacob-long** [comment]
That does seem sensible. I hesitate to say this in public (mostly joking), but I suppose I ran into this issue because I am not a full-blown adopter of tidyverse idioms, especially `select`. But I think even if someone doesn't speak fluent tidyverse, `skim(df, whatever_vector_you_want, plus_another)` would be easy to remember after seeing a single example of it.

**Comment by elinw** [comment]
I have been looking at this today and I'm wondering if rather than reinvent the tidyverse wheel, which is already available in all its glory and with helpers via piping (or inside the skim.df() data set definiton), we should just focus on the very specific case of one variable.   Anything more than that we can say you have to process your own subsetting using the methods of your choice, whether tidyverse, base R or something else.

If we did that, we'd then be back to the various other thoughts. That is

- create skim.vector
- create a print method for skim_v
- coerce anything that is not a data frame into a data frame.
- other

It seems to me that if we created skim.vector() we could achieve the goal of making skim_v internal and use skim.vector as a way to access the internal function. Print.skim_v on the other hand prevents us from making that function internal.  

That said, the question then is whether the output of skim.vector() should be a vector or a data frame. It feels weird to me to put in a vector and get out a skim_df data frame but a data frame would mean we could keep piping and subsequently use tidyverse verbs plus it would easily be able to use print.skim_df.  On the other hand returning a vector we would be type consistent but would have to make print.skim_vector.  

Thoughts?

**Comment by elinw** [comment]
So I have been looking at this and trying out some code ideas.  The first issue is that since there is not a class `vector` which rules out an explicit S3 `skim.vector()` method.  There would need to be a separate method for each type currently handled  by skim_v.  
So, that really means that as a practical matter it needs to be handled with skim.default. 

Unfortunately, in addition, `is.vector()` does not really work as expected.   There can't be a simple `skim.vector()` method.  There is not actually a 100%  certain way to discover if something is what skim means by a vector, which is, at base, anything that can make up a column in a data frame.  That is, the vectors may have attributes (e.g. ts vectors) or may not be from the standard list of atomic types listed in ?is.vector.  

This is what I came up with.

```
skim.default <-function(.data){
  if (!is.atomic(.data) | !is.null(dim(.data))[1]){
    return(message("No skim method exists for class ", class(.data), "."))
  }
  skimmed <- skim_v(.data)
  skimmed$variable <- deparse(substitute(.data))
  structure(skimmed, class = c("skim_vector", class(skimmed)) )
  
}
```
The advantage of this is providing a better error message if an uncovered object class is submitted. The disadvantage is that some matrix-like data will still get through.  Basically, objects of class `matrix` are, like vectors, atomic.  However they have dimensions while vectors do not. So they will be successfully excluded and get the return message.  That said, objects of other types, such as `dist` will still get through. That is because they are really like a cross between a vector and a matrix.     I actually am okay with this, since `skim()` will let the user know what the type is and also will provide reasonable statistics (n, n_complete etc).    So in that sense the object is skimmed.

For the print this is what I have:

```
print.skim_vector <- function(x, ...) {
  cat("Skim summary statistics\n")

  skim_print(x, groups = as.null())

}

```

@michaelquinn32 @jacob-long any thoughts?

**Comment by elinw** [comment]
Also I was thinking that if we did this we could definitely make skim_v() not be exported which would, however, be a backward compatibility break. Still if anyone is using it they can easily access  it with `:::`.

**Comment by jacob-long** [comment]
Good catch @elinw, the lack of an actual vector class is something I never even thought about despite working with these types of R objects all the time! 

I do think the default method you've proposed makes good sense and can potentially cover other data types that skim couldn't otherwise recognize.

As for what to do with skim_v, I'm not sure whether you should un-export it. This is one of those times in which it's a bit of a shame that we package developers don't have any data on how the packages are used. Maybe nobody ever uses skim_v and wouldn't miss it. A more conservative strategy is to create some duplicate non-exported function, `skimv` or some such, and leave skim_v exported as it is but add a deprecated warning and eventually remove it after some time has passed.

**Comment by elinw** [comment]
https://github.com/ropenscilabs/skimr/pull/182
Yes we could do deprecation of the external version that's worth thinking about, I could definitely see that people could be using it.
Meanwhile here is a PR. https://github.com/ropenscilabs/skimr/pull/182

**Comment by AmeliaMN** [comment]
I know I'm jumping in late, but I think supporting the tidyverse-like `skim(df, whatever_vector_you_want)` (or, `df %>% skim(whatever_vector_you_want)` as it would most naturally come from my fingers) is the right approach. Since `skimr` works with the pipe, with grouped data frames, etc., `skim(df$whatever_vector_you_want)` would seem like a strange use of syntax. I agree with @elinw that we don't need to support the case of more than one variable.

**Comment by elinw** [comment]
So I played a bit with `tidyselect` which is both simpler and more complicated than it seems.

If I do this:

```
skim.data.frame <- function(.data, .vars = NULL,  ...) {
  if (is.null(.vars[1])){
    .vars <- tidyselect::everything(.data)
  }
  selected <- tidyselect::vars_select(names(.data), .vars)
  .data <- .data[selected]
  rows <- purrr::map(.data, skim_v)
  combined <- dplyr::bind_rows(rows, .id = "variable")
  structure(combined, class = c("skim_df", class(combined)),
            data_rows = nrow(.data), data_cols = ncol(.data))
}

```
`skim(iris)` still works, `skim(iris, 1:3)` works and `skim(iris, "Petal.Width")` works but  `skim(iris, Petal.Width)` fails because it does not recognize Petal.Width as an object (since it isn't one).   Doing `skim(iris, iris$Petal.Width)` would require adding handling for objects.  `skim(iris, tidyselect::vars_select(names(iris), tidyselect::starts_with("Petal")))`  works but seems insane, though it could be documented  because advanced users could use it.  Basically, though, that is generated a set of quoted names.   I don't see a simple way to get around quoting the name.  

 So my thinking right now is that handling specified column(s) from a data frame is a different problem from handling vectors that are really vectors.  I think that the way that I've proposed to handle actual vectors is right and consistent with the idea of an S3 method, but I also think we should handle the subsetting data frames as part of skim.data.frame.  That said, I don't want to imply we are going to provide support for a million different kinds of objects, but I think vector and data.frame would support most objects that would be used in intro stats and most of the data in the datasets package (except for matrixes and xtab tables ).   

@michaelquinn32 @AmeliaMN @jacob-long

**Comment by elinw** [comment]
On the other hand this

```
skim.data.frame <- function(.data, ... ) {

  .dots <- pryr::dots(...)
  .vars <- as.character(.dots)

  if (length(.vars) == 0){
    .vars <- tidyselect::everything(.data)
  }
  selected <- tidyselect::vars_select(names(.data), .vars)

  .data <- .data[selected]
  rows <- purrr::map(.data, skim_v)
  combined <- dplyr::bind_rows(rows, .id = "variable")
  structure(combined, class = c("skim_df", class(combined)),
            data_rows = nrow(.data), data_cols = ncol(.data))
}
```

Supports `skim(iris)` and `skim(iris, Petal.Length, Petal.Width)` but not numeric selectors like 1:3  or the other select helpers.  But it is short and to the point.  Perhaps for those we should insist people use `select()` first?

**Comment by elinw** [comment]
https://github.com/ropenscilabs/skimr/pull/195  Okay here is a PR using tidyselect for selecting columns via unquoted names.  But it doesn't support other tidy select methods, which would have to be used to prefilter.  So I think we should merge both this and the other one for vectors as vectors (and to create default handling).

**Comment by elinw** [comment]
@michaelquinn32 @AmeliaMN @jacob-long  I think I'm going to merge #195  into develop and then if someone has a way to make it better (like supporting additional select helpers) that can be an enhancement.  I think it would be great to to a merge of develop into master soon, up the version number and may actually do a CRAN submission so please give feed back based on that.

**Comment by elinw** [comment]
And merged the vector handling also #194
